File: Algorithms/DLX/Java/DLX.java
/**
 * Dancing Links (DLX) Sudoku Solver - Java Implementation
 * Port of Algorithm X with Dancing Links from C reference
 *
 * Algorithm X: Exact cover problem solver using Knuth's Dancing Links technique
 * Sudoku as Exact Cover: 324 constraints (81 cells, 81 rows, 81 cols, 81 boxes)
 * Expected iterations for Matrix 1: 43
 */

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class DLX {
    // ========================================================================
    // DATA STRUCTURES
    // ========================================================================

    /**
     * DlxNode - represents a 1 in the exact cover matrix
     * Circular doubly-linked in 4 directions (up, down, left, right)
     */
    static class DlxNode {
        DlxNode up, down, left, right;
        DlxColumn column;  // Pointer to column header
        int rowId;         // ID of the row this node belongs to

        DlxNode() {
            up = down = left = right = this;
        }
    }

    /**
     * DlxColumn - column header (extends DlxNode)
     * Tracks number of nodes in column and provides name for debugging
     */
    static class DlxColumn extends DlxNode {
        int size;       // Number of nodes in this column
        String name;    // For debugging/identification

        DlxColumn(String name) {
            super();
            this.name = name;
            this.size = 0;
            this.column = this;
        }
    }

    /**
     * RowInfo - metadata to map DLX rows back to Sudoku placements
     */
    static class RowInfo {
        int row, col, num;

        RowInfo(int row, int col, int num) {
            this.row = row;
            this.col = col;
            this.num = num;
        }
    }

    // ========================================================================
    // GLOBAL STATE
    // ========================================================================

    private static int[][] puzzle = new int[9][9];
    private static int[][] solutionGrid = new int[9][9];
    private static int dlxIterations = 0;

    private static DlxColumn root;
    private static DlxColumn[] columns = new DlxColumn[324];
    private static RowInfo[] rowInfo = new RowInfo[729];
    private static DlxNode[] rowStarts = new DlxNode[729];

    // ========================================================================
    // CONSTRAINT COLUMN CALCULATIONS
    // ========================================================================

    /**
     * Calculate constraint column indices for Sudoku cell (r,c) with value n:
     * - Position constraint: (r,c) must be filled -> column: r*9 + c
     * - Row constraint: row r must have number n -> column: 81 + r*9 + (n-1)
     * - Column constraint: column c must have number n -> column: 162 + c*9 + (n-1)
     * - Box constraint: box b must have number n -> column: 243 + b*9 + (n-1)
     */

    private static int getPositionCol(int r, int c) {
        return r * 9 + c;
    }

    private static int getRowCol(int r, int n) {
        return 81 + r * 9 + (n - 1);
    }

    private static int getColCol(int c, int n) {
        return 162 + c * 9 + (n - 1);
    }

    private static int getBoxCol(int r, int c, int n) {
        int box = (r / 3) * 3 + (c / 3);
        return 243 + box * 9 + (n - 1);
    }

    // ========================================================================
    // DLX MATRIX INITIALIZATION
    // ========================================================================

    private static void initDlxMatrix() {
        // Create root column
        root = new DlxColumn("root");
        root.rowId = -1;

        // Create 324 column headers
        for (int i = 0; i < 324; i++) {
            columns[i] = new DlxColumn("C" + i);
            columns[i].rowId = -1;

            // Link into header list (insert at end, before root)
            columns[i].left = root.left;
            columns[i].right = root;
            root.left.right = columns[i];
            root.left = columns[i];
        }
    }

    private static DlxNode addNode(DlxColumn col, int rowId) {
        DlxNode node = new DlxNode();
        node.column = col;
        node.rowId = rowId;

        // Insert at end of column's circular list (before column header)
        node.down = col;
        node.up = col.up;
        col.up.down = node;
        col.up = node;
        col.size++;

        return node;
    }

    /**
     * Build a DLX row for Sudoku cell (r,c) with value n
     * Creates 4 nodes (one for each constraint) linked horizontally
     */
    private static void buildDlxRow(int r, int c, int n, int rowId) {
        // Store row metadata
        rowInfo[rowId] = new RowInfo(r, c, n);

        // Create nodes for the 4 constraints
        DlxNode n1 = addNode(columns[getPositionCol(r, c)], rowId);
        DlxNode n2 = addNode(columns[getRowCol(r, n)], rowId);
        DlxNode n3 = addNode(columns[getColCol(c, n)], rowId);
        DlxNode n4 = addNode(columns[getBoxCol(r, c, n)], rowId);

        // Link nodes horizontally in circular list
        n1.right = n2;
        n2.right = n3;
        n3.right = n4;
        n4.right = n1;

        n1.left = n4;
        n2.left = n1;
        n3.left = n2;
        n4.left = n3;

        // Store first node for this row
        rowStarts[rowId] = n1;
    }

    /**
     * Build complete DLX matrix from puzzle
     * For clue cells: create 1 row with that value
     * For empty cells: create 9 rows (one for each possible value)
     */
    private static void buildDlxMatrixFromPuzzle() {
        int rowId = 0;

        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                if (puzzle[r][c] != 0) {
                    // Cell has a clue - create only one row
                    buildDlxRow(r, c, puzzle[r][c], rowId++);
                } else {
                    // Cell is empty - create rows for all possible values
                    for (int n = 1; n <= 9; n++) {
                        buildDlxRow(r, c, n, rowId++);
                    }
                }
            }
        }
    }

    /**
     * Cover clues: pre-select rows corresponding to given values
     * Covers all 4 constraints for each clue
     */
    private static void coverClues() {
        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                if (puzzle[r][c] != 0) {
                    int n = puzzle[r][c];

                    // Find the row for this clue
                    for (int rowId = 0; rowId < 729; rowId++) {
                        if (rowStarts[rowId] != null &&
                            rowInfo[rowId].row == r &&
                            rowInfo[rowId].col == c &&
                            rowInfo[rowId].num == n) {

                            // Cover all columns in this row
                            DlxNode node = rowStarts[rowId];
                            DlxNode curr = node;
                            do {
                                coverColumn(curr.column);
                                curr = curr.right;
                            } while (curr != node);
                            break;
                        }
                    }
                }
            }
        }
    }

    // ========================================================================
    // DANCING LINKS OPERATIONS
    // ========================================================================

    /**
     * Cover a column in the DLX matrix
     * Remove column header and all rows containing this column
     * This is the "dancing" part - links can be restored exactly
     */
    private static void coverColumn(DlxColumn c) {
        // Remove column header from header list
        c.right.left = c.left;
        c.left.right = c.right;

        // For each row in this column
        for (DlxNode rowNode = c.down; rowNode != c; rowNode = rowNode.down) {
            // For each node in this row (excluding the column itself)
            for (DlxNode rightNode = rowNode.right; rightNode != rowNode; rightNode = rightNode.right) {
                // Remove this node from its column
                rightNode.down.up = rightNode.up;
                rightNode.up.down = rightNode.down;
                rightNode.column.size--;
            }
        }
    }

    /**
     * Uncover a column (exact reverse of cover)
     * Must be called in reverse order of cover for correctness
     */
    private static void uncoverColumn(DlxColumn c) {
        // For each row in this column (in reverse order)
        for (DlxNode rowNode = c.up; rowNode != c; rowNode = rowNode.up) {
            // For each node in this row (in reverse order)
            for (DlxNode leftNode = rowNode.left; leftNode != rowNode; leftNode = leftNode.left) {
                // Restore this node to its column
                leftNode.column.size++;
                leftNode.down.up = leftNode;
                leftNode.up.down = leftNode;
            }
        }

        // Restore column header to header list
        c.right.left = c;
        c.left.right = c;
    }

    /**
     * Choose column with minimum size (Knuth's S heuristic)
     * Minimizes branching factor
     */
    private static DlxColumn chooseColumn() {
        DlxColumn best = null;
        int minSize = Integer.MAX_VALUE;

        for (DlxNode colNode = root.right; colNode != root; colNode = colNode.right) {
            DlxColumn col = (DlxColumn) colNode;
            if (col.size < minSize) {
                minSize = col.size;
                best = col;
            }
        }

        return best;
    }

    /**
     * DLX Search - Algorithm X with Dancing Links
     * Recursive backtracking search for exact cover
     * Returns true if solution found, false otherwise
     */
    private static boolean dlxSearch(int k, int[] solution) {
        dlxIterations++;  // Count every search call

        // If matrix is empty, we found a solution
        if (root.right == root) {
            return true;
        }

        // Choose column with minimum size
        DlxColumn col = chooseColumn();

        // If column has no rows, no solution possible
        if (col.size == 0) {
            return false;
        }

        // Cover this column
        coverColumn(col);

        // Try each row in this column
        for (DlxNode rowNode = col.down; rowNode != col; rowNode = rowNode.down) {
            // Add row to partial solution
            solution[k] = rowNode.rowId;

            // Cover all other columns in this row
            for (DlxNode rightNode = rowNode.right; rightNode != rowNode; rightNode = rightNode.right) {
                coverColumn(rightNode.column);
            }

            // Recurse
            if (dlxSearch(k + 1, solution)) {
                return true;  // Solution found
            }

            // Backtrack: uncover all columns in this row (in reverse order)
            for (DlxNode leftNode = rowNode.left; leftNode != rowNode; leftNode = leftNode.left) {
                uncoverColumn(leftNode.column);
            }
        }

        // Uncover column
        uncoverColumn(col);

        return false;  // No solution found
    }

    // ========================================================================
    // SOLUTION EXTRACTION
    // ========================================================================

    /**
     * Extract solution from DLX and populate solutionGrid
     */
    private static void extractSolution(int[] solution, int solutionLen) {
        // Initialize solution grid with original puzzle (includes clues)
        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                solutionGrid[r][c] = puzzle[r][c];
            }
        }

        // Fill in solution values
        for (int i = 0; i < solutionLen; i++) {
            int rowId = solution[i];
            if (rowId >= 0 && rowId < 729 && rowInfo[rowId] != null) {
                RowInfo info = rowInfo[rowId];
                solutionGrid[info.row][info.col] = info.num;
            }
        }
    }

    // ========================================================================
    // PUZZLE I/O
    // ========================================================================

    private static void printPuzzle(int[][] grid) {
        System.out.println("\nPuzzle:");
        for (int r = 0; r < 9; r++) {
            StringBuilder sb = new StringBuilder();
            for (int c = 0; c < 9; c++) {
                sb.append(grid[r][c]).append(" ");
            }
            System.out.println(sb.toString());
        }
    }

    private static void readMatrixFile(String filename) {
        // Normalize path for output (match C format)
        String displayPath = filename;
        if (filename.startsWith("/app/Matrices/")) {
            displayPath = "../" + filename.substring(5);
        }
        System.out.println(displayPath);

        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            int lineCount = 0;

            while ((line = br.readLine()) != null && lineCount < 9) {
                line = line.trim();
                // Skip comments and empty lines
                if (line.isEmpty() || line.startsWith("#")) continue;

                // Parse 9 integers from line
                String[] parts = line.split("\\s+");
                if (parts.length == 9) {
                    StringBuilder sb = new StringBuilder();
                    for (int j = 0; j < 9; j++) {
                        puzzle[lineCount][j] = Integer.parseInt(parts[j]);
                        sb.append(puzzle[lineCount][j]).append(" ");
                    }
                    System.out.println(sb.toString());
                    lineCount++;
                }
            }
        } catch (IOException | NumberFormatException e) {
            System.err.println("Error reading file: " + e.getMessage());
            System.exit(1);
        }
    }

    // ========================================================================
    // MAIN ENTRYPOINT
    // ========================================================================

    public static void main(String[] args) {
        long startTime = System.nanoTime();

        // Process each .matrix file from command line
        for (String arg : args) {
            if (!arg.endsWith(".matrix")) continue;

            // Reset puzzle
            puzzle = new int[9][9];
            readMatrixFile(arg);
            printPuzzle(puzzle);

            // Initialize DLX matrix
            initDlxMatrix();

            // Build matrix from puzzle
            buildDlxMatrixFromPuzzle();

            // Cover pre-filled clues
            coverClues();

            // Solve using DLX
            dlxIterations = 0;
            int[] solution = new int[81];
            boolean result = dlxSearch(0, solution);

            if (result) {
                extractSolution(solution, 81);
                printPuzzle(solutionGrid);
                System.out.println("\nSolved in Iterations=" + dlxIterations + "\n");
            } else {
                System.out.println("\nNo solution found\n");
            }
        }

        double elapsed = (System.nanoTime() - startTime) / 1_000_000_000.0;
        System.out.printf("Seconds to process %.3f%n", elapsed);
    }
}

File: Algorithms/DLX/Kotlin/DLX.kt
/**
 * Sudoku Solver - DLX (Dancing Links) Implementation in Kotlin
 *
 * Algorithm X with Dancing Links - exact cover approach to Sudoku.
 * Maintains circular doubly-linked list structure for efficient constraint handling.
 *
 * Expected iterations for Matrix 1: 43
 */

import java.io.File

// DLX Node structure
open class DlxNode(
    var left: DlxNode? = null,
    var right: DlxNode? = null,
    var up: DlxNode? = null,
    var down: DlxNode? = null,
    var column: DlxColumn? = null,
    var rowId: Int = -1
)

// DLX Column header (inherits from node)
class DlxColumn(
    val name: String,
    var size: Int = 0
) : DlxNode()

// Row metadata to map DLX rows back to Sudoku (row, col, num)
data class RowInfo(val row: Int, val col: Int, val num: Int)

// Global variables
var iterations = 0
val puzzle = Array(9) { IntArray(9) }
val solutionGrid = Array(9) { IntArray(9) }
lateinit var root: DlxColumn
val columns = Array<DlxColumn?>(324) { null }
val nodes = mutableListOf<DlxNode>()
val rowInfo = Array<RowInfo?>(729) { null }
val rowStarts = Array<DlxNode?>(729) { null }

fun main(args: Array<String>) {
    val startTime = System.nanoTime()

    for (arg in args) {
        if (!arg.endsWith(".matrix")) continue

        // Reset puzzle
        for (i in 0 until 9) {
            for (j in 0 until 9) {
                puzzle[i][j] = 0
            }
        }

        readMatrixFile(arg)
        printPuzzle(puzzle)

        // Initialize DLX matrix
        initDlxMatrix()

        // Build matrix from puzzle
        buildDlxMatrixFromPuzzle()

        // Cover pre-filled clues
        coverClues()

        // Solve using DLX
        iterations = 0
        val solution = IntArray(81)
        val result = dlxSearch(root, 0, solution)

        if (result) {
            extractSolution(solution, 81)
            printPuzzle(solutionGrid)
            println("\nSolved in Iterations=$iterations\n")
        } else {
            println("\nNo solution found\n")
        }
    }

    val elapsed = (System.nanoTime() - startTime) / 1_000_000_000.0
    println("Seconds to process %.3f".format(elapsed))
}

fun readMatrixFile(filename: String) {
    // Normalize path for output (match C format)
    var displayPath = filename
    if (filename.startsWith("/app/Matrices/")) {
        displayPath = "../" + filename.substring(5)  // Skip "/app/" to get "Matrices/..."
    }
    println(displayPath)

    val file = File(filename)
    var lineCount = 0

    file.forEachLine { line ->
        if (lineCount >= 9) return@forEachLine
        val trimmed = line.trim()
        // Skip comments and empty lines
        if (trimmed.isEmpty() || trimmed.startsWith("#")) return@forEachLine

        // Parse 9 integers from line
        val parts = trimmed.split(Regex("\\s+"))
        if (parts.size == 9) {
            val sb = StringBuilder()
            for (j in 0 until 9) {
                puzzle[lineCount][j] = parts[j].toInt()
                sb.append(puzzle[lineCount][j]).append(" ")
            }
            println(sb.toString())
            lineCount++
        }
    }
}

fun printPuzzle(grid: Array<IntArray>) {
    println("\nPuzzle:")
    for (row in 0 until 9) {
        val sb = StringBuilder()
        for (col in 0 until 9) {
            sb.append(grid[row][col]).append(" ")
        }
        println(sb.toString())
    }
}

// Calculate constraint column indices
fun getPositionCol(r: Int, c: Int): Int = r * 9 + c
fun getRowCol(r: Int, n: Int): Int = 81 + r * 9 + (n - 1)
fun getColCol(c: Int, n: Int): Int = 162 + c * 9 + (n - 1)
fun getBoxCol(r: Int, c: Int, n: Int): Int {
    val box = (r / 3) * 3 + (c / 3)
    return 243 + box * 9 + (n - 1)
}

// Initialize DLX matrix structure
fun initDlxMatrix() {
    // Allocate root column
    root = DlxColumn("root")
    root.left = root
    root.right = root
    root.up = root
    root.down = root
    root.column = root

    // Allocate 324 column headers
    for (i in 0 until 324) {
        columns[i] = DlxColumn("C$i")
        val col = columns[i]!!

        // Initialize as circular list
        col.up = col
        col.down = col
        col.column = col

        // Link into header list
        col.left = root.left
        col.right = root
        root.left!!.right = col
        root.left = col
    }

    nodes.clear()
}

// Add a node to the DLX matrix
fun addNode(col: DlxColumn, rowId: Int): DlxNode {
    val node = DlxNode()
    node.column = col
    node.rowId = rowId

    // Insert at end of column's circular list
    node.down = col
    node.up = col.up
    col.up!!.down = node
    col.up = node
    col.size++

    nodes.add(node)
    return node
}

// Build a DLX row for Sudoku cell (r,c) with value n
fun buildDlxRow(r: Int, c: Int, n: Int, rowId: Int) {
    // Store row metadata
    rowInfo[rowId] = RowInfo(r, c, n)

    // Create nodes for the 4 constraints
    val n1 = addNode(columns[getPositionCol(r, c)]!!, rowId)
    val n2 = addNode(columns[getRowCol(r, n)]!!, rowId)
    val n3 = addNode(columns[getColCol(c, n)]!!, rowId)
    val n4 = addNode(columns[getBoxCol(r, c, n)]!!, rowId)

    // Link nodes horizontally in circular list
    n1.right = n2
    n2.right = n3
    n3.right = n4
    n4.right = n1

    n1.left = n4
    n2.left = n1
    n3.left = n2
    n4.left = n3

    // Store first node for this row
    rowStarts[rowId] = n1
}

// Build the complete DLX matrix from the puzzle
fun buildDlxMatrixFromPuzzle() {
    var rowId = 0

    for (r in 0 until 9) {
        for (c in 0 until 9) {
            if (puzzle[r][c] != 0) {
                // Cell has a clue - create only one row for that value
                buildDlxRow(r, c, puzzle[r][c], rowId++)
            } else {
                // Cell is empty - create rows for all possible values
                for (n in 1..9) {
                    buildDlxRow(r, c, n, rowId++)
                }
            }
        }
    }
}

// Cover a column in the DLX matrix
fun coverColumn(c: DlxColumn) {
    // Remove column header from the header list
    c.right!!.left = c.left
    c.left!!.right = c.right

    // For each row in this column
    var rowNode = c.down
    while (rowNode != c) {
        // For each node in this row (excluding the column itself)
        var rightNode = rowNode!!.right
        while (rightNode != rowNode) {
            // Remove this node from its column
            rightNode!!.down!!.up = rightNode.up
            rightNode.up!!.down = rightNode.down
            rightNode.column!!.size--
            rightNode = rightNode.right
        }
        rowNode = rowNode.down
    }
}

// Uncover a column (exact reverse of cover)
fun uncoverColumn(c: DlxColumn) {
    // For each row in this column (in reverse order)
    var rowNode = c.up
    while (rowNode != c) {
        // For each node in this row (in reverse order)
        var leftNode = rowNode!!.left
        while (leftNode != rowNode) {
            // Restore this node to its column
            leftNode!!.column!!.size++
            leftNode.down!!.up = leftNode
            leftNode.up!!.down = leftNode
            leftNode = leftNode.left
        }
        rowNode = rowNode.up
    }

    // Restore column header to the header list
    c.right!!.left = c
    c.left!!.right = c
}

// Choose column with minimum size (Knuth's S heuristic)
fun chooseColumn(root: DlxColumn): DlxColumn? {
    var best: DlxColumn? = null
    var minSize = Int.MAX_VALUE

    var colNode = root.right
    while (colNode != root) {
        val col = colNode as DlxColumn
        if (col.size < minSize) {
            minSize = col.size
            best = col
        }
        colNode = colNode.right
    }

    return best
}

// DLX Search - Algorithm X with Dancing Links
fun dlxSearch(root: DlxColumn, k: Int, solution: IntArray): Boolean {
    iterations++  // Count every search call (analogous to brute-force iterations)

    // If matrix is empty, we found a solution
    if (root.right == root) {
        return true
    }

    // Choose column with minimum size
    val col = chooseColumn(root) ?: return false

    // If column has no rows, no solution possible
    if (col.size == 0) {
        return false
    }

    // Cover this column
    coverColumn(col)

    // Try each row in this column
    var rowNode = col.down
    while (rowNode != col) {
        // Add row to partial solution
        solution[k] = rowNode!!.rowId

        // Cover all other columns in this row
        var rightNode = rowNode.right
        while (rightNode != rowNode) {
            coverColumn(rightNode!!.column!!)
            rightNode = rightNode.right
        }

        // Recurse
        if (dlxSearch(root, k + 1, solution)) {
            return true  // Solution found
        }

        // Backtrack: uncover all columns in this row
        var leftNode = rowNode.left
        while (leftNode != rowNode) {
            uncoverColumn(leftNode!!.column!!)
            leftNode = leftNode.left
        }

        rowNode = rowNode.down
    }

    // Uncover column
    uncoverColumn(col)

    return false  // No solution found
}

// Extract solution from DLX and populate solutionGrid
fun extractSolution(solution: IntArray, solutionLen: Int) {
    // Initialize solution grid - start with the original puzzle (includes clues)
    for (r in 0 until 9) {
        for (c in 0 until 9) {
            solutionGrid[r][c] = puzzle[r][c]
        }
    }

    // Each solution entry is a row_id
    for (i in 0 until solutionLen) {
        val rowId = solution[i]
        if (rowId >= 0 && rowId < 729) {
            val info = rowInfo[rowId]
            if (info != null) {
                solutionGrid[info.row][info.col] = info.num
            }
        }
    }
}

// Cover given clues (pre-selected rows)
fun coverClues() {
    for (r in 0 until 9) {
        for (c in 0 until 9) {
            if (puzzle[r][c] != 0) {
                val n = puzzle[r][c]

                // Find the row for this clue
                for (rowId in 0 until 729) {
                    val info = rowInfo[rowId]
                    if (rowStarts[rowId] != null &&
                        info != null &&
                        info.row == r &&
                        info.col == c &&
                        info.num == n) {

                        // Cover all columns in this row
                        val node = rowStarts[rowId]!!
                        var curr = node
                        do {
                            coverColumn(curr.column!!)
                            curr = curr.right!!
                        } while (curr != node)
                        break
                    }
                }
            }
        }
    }
}
